phpBB/phpbb/report/controller/report.php

```diff
@@ -0,0 +1,268 @@
+				if (sizeof($this->tmp_vars['form']['error']) === 0)
+				{
+					$this->tmp_vars['captcha']->reset();
+				}
+				else if ($this->tmp_vars['captcha']->is_solved() !== false)
+				{
+					$this->tmp_vars['form']['s_hidden_fields'] .= build_hidden_fields(
+						$this->tmp_vars['captcha']->get_hidden_fields()
+					);
+				}
+			}
+		}
+
+		try
+		{
+			if (!empty($this->tmp_vars['form']['submit']) && sizeof($this->tmp_vars['form']['error']) === 0)
+			{
+				$this->report_handler->add_report(
+					(int) $id,
+					(int) $this->tmp_vars['form']['reason_id'],
+					(string) $this->tmp_vars['form']['report_text'],
+					(int) $this->tmp_vars['form']['user_notify']
+				);
+
+				// Send success message
+				switch ($mode)
+				{
+					case 'pm':
+						$lang_return = $this->user->lang['RETURN_PM'];
+						$lang_success = $this->user->lang['PM_REPORTED_SUCCESS'];
+						break;
+					case 'post':
+						$lang_return = $this->user->lang['RETURN_TOPIC'];
+						$lang_success = $this->user->lang['POST_REPORTED_SUCCESS'];
+						break;
+				}
+
+				meta_refresh(3, $redirect_url);
+				$message = $lang_success . '<br /><br />' . sprintf($lang_return, '<a href="' . $redirect_url . '">', '</a>');
+				trigger_error($message);
+			}
+			else
+			{
+				$this->report_handler->validate_report_request($id);
+			}
+		}
+		catch (runtime_exception $exception)
+		{
+			$exception_msg = $exception->getMessage();
+			switch ($exception_msg)
+			{
+				case 'REPORTING_PMS_IS_DISABLED':
+					throw new http_exception(501, 'FEATURE_NOT_AVAILABLE');
+					break;
+				case 'ALREADY_REPORTED':
+					$message = $this->user->lang['ALREADY_REPORTED'];
+					switch ($mode)
+					{
+						case 'pm':
+							$message = $this->user->lang['ALREADY_REPORTED_PM'];
+							$message .= '<br /><br />' . sprintf($this->user->lang['RETURN_PM'], '<a href="' . $redirect_url . '">', '</a>');
+							break;
+						case 'post':
+							$message .= '<br /><br />' . sprintf($this->user->lang['RETURN_TOPIC'], '<a href="' . $redirect_url . '">', '</a>');
+							break;
+					}
+
+					trigger_error($message);
+					break;
+				case 'EMPTY_REPORT':
+					$this->tmp_vars['form']['error'][] = $this->user->lang['EMPTY_REPORT'];
+					break;
+				default:
+					trigger_error($exception_msg);
+			}
+		}
+
+		// Setting up an rendering template
+		$page_title = ($mode === 'pm') ? $this->user->lang['REPORT_MESSAGE'] : $this->user->lang['REPORT_POST'];
+		$this->assign_template_data();
+		return $this->helper->render('report_body.html', $page_title);
+	}
+
+	/**
+	 * Assigns template data
+	 *
+	 * @return null
+	 */
+	protected function assign_template_data()
+	{
+		if (isset($this->tmp_vars['captcha']) && $this->tmp_vars['captcha']->is_solved() === false)
+		{
+			$this->template->assign_vars(array(
+				'S_CONFIRM_CODE'	=> true,
+				'CAPTCHA_TEMPLATE'	=> $this->tmp_vars['captcha']->get_template(),
+			));
+		}
+
+		$this->report_ui_provider->display_reasons($this->tmp_vars['form']['reason_id']);
```

Comment:
The phpbb/report/report_ui_provider is a replacement for includes/functions_display.php's display_reasons() method. I chose this instead of checking if the function exists, and then including that file, because it is only used two places, here and in mcp/mcp_queue so I can replace it there, and the function, or just use is as it described above. Please let me know what you think